### ISSUE TITLE
SRV strict check (RFC 2782) breaks integration with tagged consul service discovery

### ISSUE THREAD

**Issue by xytis** [docker/docker#25534]
**Output of `docker version`:**

```
Docker version 1.12.0, build 8eab29e
```

**Output of `docker info`:**

```
Containers: 8
 Running: 8
 Paused: 0
 Stopped: 0
Images: 7
Server Version: 1.12.0
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: dnet bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 125.6 GiB
Name: docker-03
ID: 4RI3:HJSF:UBWG:MNAO:HQJV:3444:SBTU:F2DV:D4OM:23LW:V3D2:P7XO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Cluster Store: consul://127.0.0.1:8500
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Reproducible anywhere.

**Steps to reproduce the issue:**
1. Create bridge network `docker network create test`
2. Launch `docker run --rm -ti --net test azukiapp/dig dig SRV google.com`

**Describe the results you received:**
Internal resolver fails to return any result. Code terminates at https://github.com/docker/docker/blame/b1ae883d3f8c35f9142966af7bf5278e7a10b268/vendor/src/github.com/docker/libnetwork/sandbox.go#L454 and emits message "invalid protocol in service, google.com."

**Describe the results you expected:**
I expect full DNS response for requested service:
```
$ dig SRV google.com         

; <<>> DiG 9.8.3-P1 <<>> SRV google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 64903
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0

;; QUESTION SECTION:
;google.com.			IN	SRV

;; AUTHORITY SECTION:
google.com.		60	IN	SOA	ns2.google.com. dns-admin.google.com. 129737281 900 900 1800 60

;; Query time: 39 msec
;; SERVER: 10.15.20.15#53(10.15.20.15)
;; WHEN: Tue Aug  9 16:02:11 2016
;; MSG SIZE  rcvd: 78
```

While the example does return no ANSWER section, in my infrastructure I use consul, which may resolve SRV requests that are formatted not according to RFC 2782, i.e. consul.service.consul (https://www.consul.io/docs/agent/dns.html)

**Comment by thaJeztah** [comment]
This looks to be an issue in libnetwork; would you be able to open an issue in the libnetwork issue tracker? https://github.com/docker/libnetwork/issues

**Comment by xytis** [comment]
Is it possible to move this issue there?

**Comment by thaJeztah** [comment]
Moving is not possible, but you can "edit" your post above, and copy/paste the content to the new issue

**Comment by xytis** [comment]
Since this is a breaking change from 1.11 to 1.12 in my usage case, could it be prioritised?

**Comment by thaJeztah** [comment]
@xytis apologies, that wasn't clear; so in 1.11, this worked without issues?

**Comment by xytis** [comment]
I haven't dug deeper, it could be related to error handling from libnetwork, but still, the strict check seems an overkill to me.

**Comment by thaJeztah** [comment]
Ah, I see it was added in https://github.com/docker/libnetwork/commit/4b6b13af2cb243cdcbaffa87d592b08fa71213b9, which is part of 1.12.

Perhaps we should keep this one open as well

ping @sanimej PTAL

**Comment by sanimej** [comment]
@xytis Yes, the current checks in the code are based on the standard definition of the SRV record query. I will look into relaxing it.

**Comment by hermanhobnob** [comment]
I'm finding this is breaking active directory auth from Jenkins. It sounds like the same sort of problem.

```
docker network create foobar
docker run --rm=true --dns <mydnsserver> --net foobar -ti ubuntu:latest /bin/bash -c 'apt-get install -y dnsutils ; dig  +short SRV _ldap._tcp.<mywindows.domain>'
```

Under 1.10.3 this works fine and the dig command returns a list of SRV records. Under 1.11 it only works when +tcp is added to the dig command line. Under 1.12 it doesn't work at all. However in all cases it does work if you remove the --net option and use the default network.

I don't have a workaround other than sticking with 1.10.3 for the time being.

**Comment by sanimej** [comment]
@hermanhobnob Issue you are seeing in 1.12 is not because of validation check for service name failing. I see an issue in forwarding the SRV queries to external servers. Will fix this for 1.12.1

In 1.11 are you seeing SRV queries failing always when using UDP ? Or does it work for sometime and then it fails ? If its latter the issue could be #22185. Its fixed in 1.12.

**Comment by hermanhobnob** [comment]
@sanimej I never saw a successful UDP SRV lookup with 1.11 in probably 10-20 attempts but it sounds like testing again once 1.12.1 is out is the best plan. Thanks for finding the forwarding issue.
